Signal/src/contact/SystemContactsFetcher.swift

```diff
@@ -0,0 +1,120 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import Contacts
+
+@objc protocol SystemContactsFetcherDelegate: class {
+    func systemContactsFetcher(_ systemContactsFetcher: SystemContactsFetcher, updatedContacts contacts: [Contact])
+}
+
+@objc
+class SystemContactsFetcher: NSObject {
+
+    private let TAG = "[SystemContactsFetcher]"
+
+    public weak var delegate: SystemContactsFetcherDelegate?
+
+    public var authorizationStatus: CNAuthorizationStatus {
+        return CNContactStore.authorizationStatus(for: CNEntityType.contacts)
+    }
+
+    public var isAuthorized: Bool {
+        guard self.authorizationStatus != .notDetermined else {
+            assertionFailure("should have called `requestOnce` before this point.")
+            Logger.error("\(TAG) should have called `requestOnce` before checking authorization status.")
+            return false
+        }
+
+        return self.authorizationStatus == .authorized
+    }
+
+    private let contactStore = CNContactStore()
+    private var systemContactsHaveBeenRequestedAtLeastOnce = false
+    private let allowedContactKeys: [CNKeyDescriptor] = [
+        CNContactFormatter.descriptorForRequiredKeys(for: .fullName),
+        CNContactThumbnailImageDataKey as CNKeyDescriptor, // TODO full image instead of thumbnail?
+        CNContactPhoneNumbersKey as CNKeyDescriptor,
+        CNContactEmailAddressesKey as CNKeyDescriptor
+    ]
+
+    public func requestOnce() {
+        AssertIsOnMainThread()
+
+        guard !systemContactsHaveBeenRequestedAtLeastOnce else {
+            Logger.debug("\(TAG) already requested system contacts")
+            return
+        }
+        systemContactsHaveBeenRequestedAtLeastOnce = true
+        self.startObservingContactChanges()
```

Comment:
> Will iOS send CNContactStoreDidChange notifications if we don't have contacts permission?

I don't know, but since the answer to that question isn't obvious, and since we *know* there aren't any meaningful things to observe until we've "requested" contacts permission, this seemed more straight forward.